extensions/notebook/src/book/remoteBookController.ts

```diff
@@ -0,0 +1,290 @@
+const msgRemoteBookDownloadError = localize('msgRemoteBookDownloadError', "Error while downloading remote Book");
+const msgRemoteBookUnpackingError = localize('msgRemoteBookUnpackingError', "Error while decompressing remote Book");
+const msgRemoteBookDirectoryError = localize('msgRemoteBookDirectoryError', "Error while creating remote book directory");
+const msgTaskName = localize('msgTaskName', "Downloading Remote Book");
+
+const msgResourceNotFound = localize('msgResourceNotFound', 'Resource not Found');
+const msgBookNotFound = localize('msgBookNotFound', 'Books not Found');
+const msgReleaseNotFound = localize('msgReleaseNotFound', 'Releases not Found');
+
+export class RemoteBookController {
+	private _book: RemoteBook;
+	private readonly assetNameRE = /([a-zA-Z0-9]+)(?:-|_)([a-zA-Z0-9.]+)(?:-|_)([a-zA-Z0-9]+).(zip|tar.gz|tgz)/;
+	constructor(public model: RemoteBookDialogModel) {
+	}
+
+	public async setRemoteBook(url: URL, remoteLocation: string, asset?: IAssets): Promise<void> {
+		if (remoteLocation === 'GitHub') {
+			this._book = new GitHubRemoteBook(url, asset.browser_download_url);
+		} else {
+			this._book = new SharedRemoteBook(url);
+		}
+		return await this._book.createLocalCopy();
+	}
+
+	public async fetchGithubReleases(url: URL): Promise<IReleases[]> {
+		let options = {
+			headers: {
+				'User-Agent': 'request'
+			}
+		};
+		return new Promise<IReleases[]>((resolve, reject) => {
+			request.get(url.href, options, (error, response, body) => {
+				if (error) {
+					return reject(error);
+				}
+
+				if (response.statusCode === 404) {
+					return reject(msgResourceNotFound);
+				}
+
+				if (response.statusCode !== 200) {
+					return reject(response.statusCode);
+				}
+				let releases = JSON.parse(body);
+				let bookReleases: IReleases[] = [];
+				if (releases) {
+					let keys = Object.keys(releases);
+					keys = keys.filter(key => {
+						let release = {} as IReleases;
+						try {
+							release.name = releases[key].name;
+							release.assets_url = new URL(releases[key].assets_url);
+						}
+						catch (error) {
+							return reject(error);
+						}
+						bookReleases.push(release);
+					});
+				}
+				if (bookReleases) {
+					this.model.releases = bookReleases;
+					resolve(bookReleases);
+				} else {
+					return reject(msgReleaseNotFound);
+				}
+			});
+		});
+	}
+
+	public async fecthListAssets(release: IReleases): Promise<IAssets[]> {
+		let format: string[] = [];
+		if (utils.getOSPlatform() === utils.Platform.Windows || utils.getOSPlatform() === utils.Platform.Mac) {
+			format = ['zip'];
+		} else {
+			format = ['tar.gz', 'tgz'];
+		}
+		let options = {
+			headers: {
+				'User-Agent': 'request'
+			}
+		};
+		return new Promise<IAssets[]>((resolve, reject) => {
+			request.get(release.assets_url.href, options, (error, response, body) => {
+				if (error) {
+					return reject(error);
+				}
+
+				if (response.statusCode === 404) {
+					return reject(msgResourceNotFound);
```

Comment:
Same here about the error objects and including the message for the error status code response.